R/survival_qp.R

```diff
@@ -0,0 +1,219 @@
+  l1 <- n
+  u1 <- n
+  
+  if(verbose) message("\tx Summed weight constraint for non-censored control group")
+  A2 <- as.integer(noncens_t & (trt == 0))
+  l2 <- n
+  u2 <- n
+  
+  if(verbose) message("\tx Restrict weights for non-censored individuals (between 1 and n)")
+  A3 <- diag(as.integer(noncens_t))
+  l3 <- rep(1, n)
+  u3 <- rep(n, n)
+  
+  if(verbose) message("\tx Restrict weights for censored individuals (equal to 0)")
+  A4 <- diag(as.integer(!noncens_t))
+  l4 <- rep(0, length(noncens_t))
+  u4 <- rep(0, length(noncens_t))
+  
+  if(verbose) message("\tx Combining constraints")
+  A <- rbind(A1, A2, A3, A4)
+  l <- c(l1, l2, l3, l4)
+  u <- c(u1, u2, u3, u4)
+  
+  if(verbose) message("\tx Removing null constraints")
+  nonNullConstrIdxs <- which(rowSums(A) != 0)
+  A <- A[nonNullConstrIdxs, ]
+  l <- l[nonNullConstrIdxs]
+  u <- u[nonNullConstrIdxs]
+
+  return(list(A = A, l = l, u = u))
+}
+
+#' Check that data is in right shape, hyparameters are feasible, and basis fns are valid
+#' @param B_X n x k basis matrix of covariates (X)
+#' @param trt Vector of treatment assignments
+#' @param times Vector of event/censoring times (see events for if event or censoring time)
+#' @param events Vector of boolean censoring indicators (whether individual was censored)
+#' @param t Time
+#' @param lambda Regularization hyper parameter
+#' @param lowlim Lower limit on weights, default 1
+#' @param uplim Upper limit on weights, default NULL
+check_data_surv <- function(B_X, trt, times, events, t, lambda, lowlim = 1, uplim = NULL) {
+  # NA checks
+  if(any(is.na(B_X))) {
+    stop("Basis matrix B_X contains NA values")
+  }
+  
+  if(any(! trt %in% c(0,1))) {
+    stop("Treatment must be (0,1)")
+  }
+  
+  if(any(! events %in% c(TRUE,FALSE))) {
+    stop("Censoring indicators must be (TRUE,FALSE)")
+  }
+  
+  if(any(is.na(times))) {
+    stop("NA values in times")
+  }
+  
+  # Group size check
+  # Vector of booleans indicating if indiv is not censored at time t
+  # TRUE: indiv is not censored at time t, FALSE: indiv is censored at time t
+  noncens_t <- ((times >= t) & (events == TRUE)) | (events == FALSE)
+  if(sum(noncens_t & (trt == 0)) == 0) {
+    stop("No non-censored individuals in treatment group at time ", t)
+  }
+  if(sum(noncens_t & (trt == 1)) == 0) {
+    stop("No non-censored individuals in control group at time ", t)
+  }
+  
+  #dimension checks
+  n <- nrow(B_X)
+  # k <- ncol(B_X)
+  
+  if (length(trt) != n) {
+    stop("Number of rows in basis matrix B_X (",
+         n, ") does not equal the length of trt (",
+         length(trt), ")")
+  }
+  
+  # hyerparameters are feasible
+  if(lambda < 0) {
+    stop("lambda must be >= 0")
+  }
+  if(lowlim > uplim) {
+    stop("Lower threshold must be lower than upper threshold")
+  }
+  if(lowlim < 0) {
+    stop("Lower threshold must be at least 0")
+  }
+  if(uplim > n) {
+    stop("Upper threshold must be at most n (",
+         n, ")")
+  }
+}
```

Comment:
Are these necessary? We might want to consider having negative weights at some point maybe.